walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
+
+        for (Event event : events) {
+            if (doIntersect(event.getAttendees(), attendees)) {
+                meetingsPoints.add(new Point(Point.Type.START, event));
+                meetingsPoints.add(new Point(Point.Type.END, event));
+            }
+        }
+        
+        return getEmptyTimeRanges(meetingsPoints, TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, duration);
+    }
+
+    /**
+     * @return all segments from firstPoint to lastPoint inclusive which are no shorter than minLength
+     *.        and do not intersect with segments from intersectingSegments
+     * @param intersectingSegments - array of Points. To put segment in this array, you should put its
+     *.                              START and END Points there.
+     * @param firstPoint - beginning of all time. Returned collection will not contain time before it.
+     * @param lastPoint - end of all time. Returned collection will not contain time after it.
+     * @param minLength - minimal length for each of the returned intervals
+     * Algorithm:
+     *.        All empty slots can start only where some segment ends and end only where some segment starts (or at
+               firstPoint or lastPoint)
+     *.        There cannot be any other points from intersectingSegments inside the empty slot - only its start end end.
+     *         Sort all points in intersectingSegments in ascending order and look at them one by one.
+     *.        Count how many segments are currently covering the point you are looking at: get a counter
+     *.        which increments every time you look at START point and decrements every time you look at END point.
+     *.        If the point is covered with 0 segments - it is a start of segment which can be one of the answer
+     *.        segments if it has appropriate length.
+     *.        So every time we look at a point, we check if the previous interval was segment-free and if so, we check
+     *.        the length of that interval to be at leash minLength.
+     */
+    private Collection<TimeRange> getEmptyTimeRanges(ArrayList<Point> intersectingSegments, int firstPoint, int lastPoint, long minLength) {
+        Collections.sort(intersectingSegments, Point.ORDER_START_BEFORE_END);
+
+        int openSegments = 0;
+        int lastTime = firstPoint;
+
+        // the array we will return
+        ArrayList<TimeRange> emptyTimeRanges = new ArrayList<>();
+
+        for (Point point : intersectingSegments) {
+            if (point.type == Point.Type.START) {
+                if (openSegments == 0 && point.time - lastTime >= minLength) {
+                    emptyTimeRanges.add(TimeRange.fromStartEnd(lastTime, point.time, false));
+                }
+                openSegments++;
+            } else {
+                openSegments--;
+            }
+            lastTime = point.time;
+        }
+
+        // check if the segment ending at lastPoint fits
+        if (lastPoint - lastTime >= minLength) {
```

Comment:
Considering that lastPoint is an inclusive end of the interval, the length of the segment is `lastPoint - lastTime + 1`